### PR TITLE
[android][web-browser] Fix WebBrowser sending 'dismiss' before opening

### LINKED ISSUE #6679: AuthSession returns dismiss result even before the browser is opened
## 🐛 Bug Report

Calling the `AuthSession.startAsync` for the first time returns the `dismiss` result immediately after calling. The browser opens correctly in the meantime, but any action done afterwards in the browser is ignored (login correctly, cancel, etc). This means that our users will go to the login screen, login successfully, but the app would understand this session as being dismissed.

In addition, but not connected to this issue directly:
* On Android, when the user cancels the login (does not open browser, clicks on `X` in browser, etc) the result sent back to us is also `dismiss` as well. This means we can not distinguish between cancel and dismiss, which prevents us to create at least a meaningful workaround for this issue.
* Calling  `AuthSession.dismiss` before starting the session crashes the app (in the Expo client)

### Environment

**IMPORTANT**:
* only happens on Android
* only happens in production mode or the built standalone app **(not reproducible in the dev mode)**
* only happens on the first call to the `AuthSession.startAsync`, successive calls bring the expected behaviour
* confirmed in both simulator and Android device

```
  Expo CLI 3.11.3 environment info:
    System:
      OS: Linux 5.0 Ubuntu 18.04.3 LTS (Bionic Beaver)
      Shell: 4.4.20 - /bin/bash
    Binaries:
      Node: 10.18.0 - /usr/bin/node
      npm: 6.13.4 - /usr/bin/npm
    npmPackages:
      @types/react: ~16.9.0 => 16.9.16 
      @types/react-native: ~0.60.23 => 0.60.25 
      @types/react-navigation: ^3.0.7 => 3.0.7 
      expo: ^36.0.2 => 36.0.2 
      react: ~16.9.0 => 16.9.0 
      react-native: https://github.com/expo/react-native/archive/sdk-36.0.0.tar.gz => 0.61.4 
      react-navigation: ^3.13.0 => 3.13.0 
    npmGlobalPackages:
      expo-cli: 3.11.3
```

### Steps to Reproduce

Simply call `AuthSession.startAsync` for the first time since the app has been loaded.

### Expected Behavior

I would expect that dismissed is not the result, especially that the browser get open and user not knowing what s going proceeds to the login page.

### Reproducible Demo
Here is the most simple demo: https://snack.expo.io/rJZ!LryxI

However this needs to be run in the production mode in order to actually get the bug. I am not sure if expo snack supports production mode.

### Extra

* I am also linking here a question from StackOverflow from another user having the same issue: https://stackoverflow.com/questions/59589158/expo-authsession-immediately-resolves-as-dismissed
* I guess this can be connected to https://github.com/expo/expo/issues/6289

### PR DESCRIPTION
**Pull request by LucaColonnello**
# Why
Fixes #6679

# How

https://github.com/expo/expo/issues/6679#issuecomment-570963717

**TL;DR**
A wrong usage of AppState causes WebBrowser to resolve with
`{ type: 'dismiss' }` before opening the view.

This is related to AppState being not initialised yet,
triggering a change before the user leaves the app view
to focus on the browser view.

# Test Plan

As this can be tested only in production (the bug does not happen in development), testing is hard to achieve.
I tested it in the app I'm building, other people (see comments in the issue) did the same, linking their forked package with the patch applied to their apps, and it seems to work fine.

### CODE REVIEW

**Comment by LucaColonnello** [comment]
I'm looking for ways to test this in E2E 🤞
Open to suggestions!

**Comment by LucaColonnello** [comment]
image
Tests are having issues in CI...

**Comment by LucaColonnello** [comment]
@mczernek I'm unfortunately blocked on this one.
Could I get some support? I don't know how to unlock this CI error.
I read the contributor guidelines document but I can't find anything relevant...

**Comment by lukmccall** [comment]
Hi @LucaColonnello, to fix the problem with SDK, you need to run `yarn build` in the `expo-web-browser` folder. 

An issue with the client is strange. Maybe rebase help.  

Moreover, please add an entry to [`CHANGELOG.md`](https://github.com/expo/expo/blob/master/CHANGELOG.md).

**Comment by LucaColonnello** [comment]
@lukmccall @mczernek @sjchmiela Can I ask for a re review here please? :)
I added the `CHANGELOG.md` and managed to do get the CI passing (all but the client 🤷‍♂)

Thanks in advance!

File: packages/expo-web-browser/src/WebBrowser.ts [hunk 1]
```diff
@@ -149,7 +149,18 @@ let _redirectHandler: ((event: RedirectEvent) => void) | null = null;
 // returns to active
 let _onWebBrowserCloseAndroid: null | (() => void) = null;
 
+// Store previous app state to check whether the listener has ever been attached
+let _previousAppState: null | string = AppState.currentState;
```
Comment (lukmccall) [inline packages/expo-web-browser/src/WebBrowser.ts]: It'll be better if we change it to `bool` and rename it. We don't have to remember the previous app state. We need information if it the first call or not.

**Review by lukmccall — CHANGES_REQUESTED** [review]
Please don't bother about failing CI job - this particular job always fails to people from community.  

I'm not sure why you have changes in `yarn.lock`. Did you rebase your branch?

File: packages/expo-web-browser/CHANGELOG.md [hunk 2]
```diff
@@ -9,3 +9,5 @@
 - Add `readerMode` and `dismissButtonStyle` (iOS) and `enableDefaultShare` (Android) flags for `WebBrowser` ([#7221](https://github.com/expo/expo/pull/7221) by [@LinusU](https://github.com/LinusU)) & [@mczernek](https://github.com/mczernek))
 
 ### 🐛 Bug fixes
+
+- Fix WebBrowser sending `dismiss` before opening ([#6743](https://github.com/expo/expo/pull/6743) by [@LucaColonnello](https://github.com/LucaColonnello))
```
Comment (lukmccall) [inline packages/expo-web-browser/CHANGELOG.md]:
```suggestion
- Fix `WebBrowser` sending `dismiss` before opening. ([#6743](https://github.com/expo/expo/pull/6743) by [@LucaColonnello](https://github.com/LucaColonnello))
```

**Review by lukmccall — APPROVED** [review]
LGTM 🥇I'll merge when you fix one more thing ;)

**Comment by LucaColonnello** [comment]
🎉
Can someone help testing this in an Android machine?
Unfortunately my machine has some issues with Android VM.

**Comment by LucaColonnello** [comment]
Hey @cruzach can you please explain your point better? Unfortunately I do not understand why it would result in dismiss.

Only thing I’m tackling here is that there’s no guarantee to have an initial event from AppState triggered ONLY when you do change focus.

The AppState could trigger the first time is currentState is called even if no change occurred.

How is this change impacting the current feature as you say?

Thanks in advance :)

**Comment by cruzach** [comment]
Sorry, I didn't mean to indicate this PR was negatively impacting the feature at all. 

Since, with a Chrome custom tab, we have no way of knowing whether the web browser was closed, or backgrounded, we [state in the docs](https://docs.expo.io/versions/latest/sdk/webbrowser/) that:
> On Android promise resolves with {type: 'opened'} if we were able to open browser.

I think we should revert back to this behavior, since resolving to `{type: 'dismiss'}` indicates that the web browser was successfully closed, which is the behavior on the iOS side:
> On iOS: if the browser is closed using `dismissBrowser` , the Promise resolves with { type: 'dismiss' }.


The best option would be to tell whether the chrome custom tab was closed or not, but I'm not sure it's possible

**Comment by brentvatne** [comment]
I can't reproduce the issue described here on SDK36 on Android.

```
import React from "react";
import { AppState, Button, StyleSheet, Text, View } from "react-native";
import * as WebBrowser from "expo-web-browser";

export default function App() {
  return (
    <View style={styles.container}>
      <Text>{AppState.currentState}</Text>
      <Button
        title="Press me"
        onPress={() => {
          WebBrowser.openAuthSessionAsync(
            "https://expo.io",
            "exp://127.0.0.1:19000"
          );
        }}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center"
  }
});
```

1) the `AppState.currentState` text is initialized to `active`, which is what is to be expected for React Native 0.61 (the docs appear to be outdated and incorrect - see https://github.com/facebook/react-native/blob/master/Libraries/AppState/AppState.js#L42
2) I added logs to the app state change function that you modified and it fired with the values I expected: first `background` and then `active` when I return to the app.

I tried this inside of Expo client and in a standalone app ([download apk here](https://exp-shell-app-assets.s3.us-west-1.amazonaws.com/android/%40notbrent/testing-focus-stuff-6bdc91ca684243ae89e727d344a9d2e6-signed.apk))

Can you explain how I can reproduce this problem on SDK36?

**Comment by thisBrian** [comment]
@brentvatne 

The reason is because you call `AppState.current` explicitly before opening the browser.
The problem will occur were you to start the auth flow directly and see if it resolves.

**Comment by LucaColonnello** [comment]
Are we ok to merge this @lukmccall? :)

**Comment by brentvatne** [comment]
@thisBrian - that doesn't make any sense -- accessing the property does not initialize it. it is initialized in the AppState constructor as i linked to. i would need more information on exactly what the problem is and how i can reproduce this before i'm comfortable merging

**Comment by thisBrian** [comment]
@brentvatne 

Apologies if it didn't make sense, let me clarify: your _code block_ where you read from AppState prior to launching the auth session has side effects that won't reproduce the bug.

[To demonstrate I forked your helloworld repo](https://github.com/nullables/hello-world-jest-expo-universal) and quickly added a simple auth flow using reddit as an example.

- [Here is the bug](https://i.imgur.com/eUdaJ8j.mp4) where the first attempt silently fails, yet subsequent attempts succeed
- [Here is the same code](https://i.imgur.com/1XklccM.mp4) but with the side effect of having read `AppState.currentState` prior to launching the browser.

The same holds true for any helper classes that use WebBrowser (such as AuthSession).

I hope that clears it up enough to merge the fix by @LucaColonnello, cheers!

----
Here's an excerpt from the aformentioned fork

```tsx
import { AuthSession } from 'expo';
import React from 'react';
import { Button, StyleSheet, View } from 'react-native';
import { getAuthorizeUrl } from './helpers';

export default function App() {
  return (
    <View style={styles.container}>
      <Button
        title="Press me"
        onPress={async () => {
          const returnUrl = 'helloworld://any';
          const authUrl = getAuthorizeUrl(
            {
              client_id: 'INSERT_YOUR_CLIENT_ID', // https://www.reddit.com/prefs/apps/ + match redirect Url with aformentioned returnUrl
              duration: 'temporary',
              redirect_uri: returnUrl,
              response_type: 'code',
              scope: ['identity'],
            },
            '.compact'
          );
          const result = await AuthSession.startAsync({
            authUrl,
            returnUrl,
          }).catch(alert);
          alert(JSON.stringify(result));
        }}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

**Comment by LucaColonnello** [comment]
@brentvatne I understand where you're coming from, it is misleading.
I'll try make it clearer.

The ReactContext `LifecycleEventListener`, used by AppState, triggers an `onHostResume` event when you add the _first ever_ listener, according to this [code](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/ReactContext.java#L200).

(read  my comment here for details of why this is important in the context of `AuthSession` https://github.com/expo/expo/issues/6679#issuecomment-570963717, although lines have changed since then in ReactNative)

You're right to say that if that is the case it should not matter as AppState constructor creates a listener as soon as an instance is created, but there's a detail you're missing.

**In the context of ReactNative, all modules are exported as getter of the `module.exports` object, requiring such modules only when and if required (most probably for performance reasons).**
Here's the [code responsible for that](https://github.com/facebook/react-native/blob/master/index.js#L294), you can see AppState being exported in such way at line 294.

Based on this, when you import AppState in your app before using `AuthSession`, you're triggering the constructor earlier.

**Comment by LucaColonnello** [comment]
@thisBrian could you tell me how to reproduce this in a virtual machine if you did this in a VM to take the videos?
I was only able to debug this in a real device (old Android), but unfortunately I don't have it with me anymore...

Thanks :)

**Comment by thisBrian** [comment]
@LucaColonnello 

Oh I had used a screen recorder for those particular videos; [however Expo has some good docs on setting up a virtual device](https://docs.expo.io/versions/latest/workflow/android-studio-emulator/).

Alternatively if you want a quick and dirty virtual Android device to run the apk, you could use [BlueStacks](https://www.bluestacks.com/) too. I have tested it with both virtual variations and they could reproduce the bug.

Apologies, if I was able to get to the office I would upload my docker image that has it all set up.

**Comment by lukmccall** [comment]
@LucaColonnello, I'm sorry it took so long. However, I think all doubts were resolved. So, we almost ready to merge it 🎉Please, just resolve conflicts and it will be good to go.

**Comment by vernondegoede** [comment]
Hey guys,

We ran into the same issue with `expo-web-browser`. When running `openAuthSessionAsync` from `expo-web-browser`, we always get returned with `{"type": "dismiss"}`. This only happens on Android.

I applied the changes from this PR using `patch-package`, but without any luck.

Just to be sure, this is what `patches/expo-web-browser+8.1.0.patch` looks like:

```
diff --git a/node_modules/expo-web-browser/build/WebBrowser.js b/node_modules/expo-web-browser/build/WebBrowser.js
index 91daac4..d0d56dc 100644
--- a/node_modules/expo-web-browser/build/WebBrowser.js
+++ b/node_modules/expo-web-browser/build/WebBrowser.js
@@ -106,7 +106,16 @@ let _redirectHandler = null;
 // Store the `resolve` function from a Promise to fire when the AppState
 // returns to active
 let _onWebBrowserCloseAndroid = null;
+// If the initial AppState.currentState is null, we assume that the first call to
+// AppState#change event is not actually triggered by a real change,
+// is triggered instead by the bridge capturing the current state
+// (https://facebook.github.io/react-native/docs/appstate#basic-usage)
+let _isAppStateAvailable = AppState.currentState !== null;
 function _onAppStateChangeAndroid(state) {
+    if (!_isAppStateAvailable) {
+        _isAppStateAvailable = true;
+        return;
+    }
     if (state === 'active' && _onWebBrowserCloseAndroid) {
         _onWebBrowserCloseAndroid();
     }
diff --git a/node_modules/expo-web-browser/src/WebBrowser.ts b/node_modules/expo-web-browser/src/WebBrowser.ts
index 37ee736..0391591 100644
--- a/node_modules/expo-web-browser/src/WebBrowser.ts
+++ b/node_modules/expo-web-browser/src/WebBrowser.ts
@@ -149,7 +149,17 @@ let _redirectHandler: ((event: RedirectEvent) => void) | null = null;
 // returns to active
 let _onWebBrowserCloseAndroid: null | (() => void) = null;
 
+// If the initial AppState.currentState is null, we assume that the first call to
+// AppState#change event is not actually triggered by a real change,
+// is triggered instead by the bridge capturing the current state
+// (https://facebook.github.io/react-native/docs/appstate#basic-usage)
+let _isAppStateAvailable: boolean = AppState.currentState !== null;
 function _onAppStateChangeAndroid(state: AppStateStatus) {
+  if (!_isAppStateAvailable) {
+    _isAppStateAvailable = true;
+    return;
+  }
+
   if (state === 'active' && _onWebBrowserCloseAndroid) {
     _onWebBrowserCloseAndroid();
   }
```

**Review by brentvatne — APPROVED** [review]
still not totally sure why this is happening but was able to repro it in a prod build of the app and not in development. thanks for the patience!

**Comment by brentvatne** [comment]
this is available in `expo-web-browser@8.1.1`

**Comment by Rohansi** [comment]
> this is available in `expo-web-browser@8.1.1`

Any way to use that in an SDK 37 app? I tried pulling it but it wasn't fixed, and `expo install`ing it only gives 8.1.0. Maybe I need to build standalone again?

**Comment by brentvatne** [comment]
if you already have expo-web-browser in your app 8.1.0 matches ~8.1.0 which is what is installed by expo install. if you remove it and then install it again you will get the latest version, or just run yarn add expo-web-browser@8.1.1

**Comment by Rohansi** [comment]
Yeah, I installed expo-web-browser@8.1.1 and rebuilt the standalone app. Still immediately get a dismiss the first time around.

Expo CLI also said `There were no new changes from the last build, you can download that build from here` which makes me think these dependencies are bundled separately from my app's code?

**Comment by simplesthing** [comment]
I am getting this issue with expo-web-browser 8.2.1

**Comment by thisBrian** [comment]
@simplesthing It seems you got your issue [resolved](https://github.com/expo/expo/issues/6679#issuecomment-637040641). btw the `./helpers` and `getAuthorizeUrl` were reddit-specific implementations for the demo to reproduce the bug.
